src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,210 @@
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    /// <summary>
+    /// Caches tokens on a per-line basis for documents based on semantic version.
+    /// </summary>
+    /// <remarks>
+    /// The cache makes the assumption that the tokens passed in are complete for the given range.
+    /// </remarks>
+    internal sealed class SemanticTokensCache
+    {
+        private const int MaxSemanticVersionPerDoc = 4;
+
+        // Nested cache mapping (URI -> (semanticVersion -> (line #s -> tokens on line)))
+        private readonly MemoryCache<DocumentUri, MemoryCache<VersionStamp, Dictionary<int, List<int>>>> _cache = new();
+
+        /// <summary>
+        /// Caches tokens on a per-line basis. If the given line has already been cached for the document
+        /// and its semantic version, the line's tokens will be ignored.
+        /// </summary>
+        /// <param name="uri">The URI associated with the passed-in tokens.</param>
+        /// <param name="semanticVersion">The document's semantic version associated with the passed-in tokens.</param>
+        /// <param name="range">The range associated with the passed-in tokens.</param>
+        /// <param name="tokens">The complete set of tokens for the passed-in range.</param>
+        internal void CacheTokens(
+            DocumentUri uri,
+            VersionStamp semanticVersion,
+            Range range,
+            int[] tokens)
+        {
+            if (!_cache.TryGetValue(uri, out var documentCache))
+            {
+                documentCache = new MemoryCache<VersionStamp, Dictionary<int, List<int>>>(sizeLimit: MaxSemanticVersionPerDoc);
+                _cache.Set(uri, documentCache);
+            }
+
+            if (!documentCache.TryGetValue(semanticVersion, out var lineToTokensDict))
+            {
+                lineToTokensDict = new Dictionary<int, List<int>>();
+                documentCache.Set(semanticVersion, lineToTokensDict);
+            }
+
+            var absoluteLine = 0;
+
+            // Cache the tokens associated with each line
+            for (var tokenIndex = 0; tokenIndex < tokens.Length; tokenIndex += 5)
+            {
+                absoluteLine += tokens[tokenIndex];
+
+                // Don't cache tokens outside of the given range
+                if (absoluteLine < range.Start.Line || absoluteLine > range.End.Line)
+                {
+                    continue;
+                }
+
+                if (lineToTokensDict.TryGetValue(absoluteLine, out _))
+                {
+                    // This line is already cached, so we can skip it
+                    continue;
+                }
+
+                var lineTokens = new List<int>();
```

Comment:
Building a new list for every line may be expensive. We may want to look into optimizations around here

Reply:
Good point, this might be expensive. I can't think of any alternatives off the top of my head though unfortunately ☹️ maybe we can brainstorm something for a follow-up PR?